cli/cmd/cluster.go

```diff
@@ -62,52 +63,80 @@ func initialize(ctxLogger *logrus.Entry) (*terraform.Executor, platform.Platform
 		ctxLogger.Fatal("No cluster configured")
 	}
 
-	// Get the configured backend for the cluster. Backend types currently supported: local, s3.
-	b, diags := getConfiguredBackend(lokoConfig)
-	if diags.HasErrors() {
-		for _, diagnostic := range diags {
-			ctxLogger.Error(diagnostic.Error())
-		}
+	// Render backend configuration.
+	var renderedBackend string
 
-		ctxLogger.Fatal("Errors found while loading cluster configuration")
-	}
+	if lokoConfig.RootConfig.Backend != nil {
+		b := createBackend(ctxLogger, lokoConfig)
 
-	// Use a local backend if no backend is configured.
-	if b == nil {
-		b = local.NewLocalBackend()
+		renderedBackend = b.String()
 	}
 
 	assetDir, err := homedir.Expand(p.Meta().AssetDir)
 	if err != nil {
 		ctxLogger.Fatalf("Error expanding path: %v", err)
 	}
 
-	// Validate backend configuration.
-	if err = b.Validate(); err != nil {
-		ctxLogger.Fatalf("Failed to validate backend configuration: %v", err)
+	ex := initializeTerraform(ctxLogger, p, renderedBackend)
+
+	return ex, p, lokoConfig, assetDir
+}
+
+// createBackend constructs a Backend based on the provided cluster config and returns a pointer to
+// it. If a backend with the provided name doesn't exist, an error is returned.
+func createBackend(logger *logrus.Entry, config *config.Config) backend.Backend {
+	bn := config.RootConfig.Backend.Name
+
+	switch bn {
+	case backend.Local:
+		bc, diags := local.NewConfig(&config.RootConfig.Backend.Config, config.EvalContext)
+		if diags.HasErrors() {
+			for _, diagnostic := range diags {
+				logger.Error(diagnostic.Error())
+			}
+
+			logger.Fatal("Errors found while loading backend configuration")
+		}
+
+		b, err := local.NewBackend(bc)
+		if err != nil {
+			logger.Fatalf("Error constructing backend: %v", err)
+		}
+
+		return b
```

Comment:
>I suggest we focus to only select this unique information rather than re-doing the same thing for each backend.

What does this mean in practice? I don't understand your suggestion.

Reply:
I suggest we only select the right structure to valuate the config into, rather than re-doing error checking and creating the backend. So we have one line per backend type, not many.

Reply:
I understand. The problem is that I wanted to do all the "work" at creation time, i.e. unmarshaling, validation etc. It doesn't make sense to me to return an uninitialized, unvalidated struct and hope that the user calls all the relevant methods on the `Backend` interface.

AFAICT, to do what you're suggesting without exposing things like `LoadConfig()` and `Validate()` on the `Backend` interface we would have to do *everything* inside `NewBackend()`, which means we would have to return "regular" Go errors inside `hcl.Diagnostics` :confounded:

That's why I've chosen to separate the HCL unmarshaling from things like custom validations and rendering the template, and that's why we have a `Config` separately from the `Backend`, and *that's* why we have to explicitly create a `Config` and a `Backend` for each backend type.

Reply:
I'm wondering if we should stop using `hcl.Diagnostics` altogether. Personally I don't like this mechanism at all and would rather deal with "regular" errors myself.

Reply:
>I'm wondering if we should stop using hcl.Diagnostics altogether. Personally I don't like this mechanism at all and would rather deal with "regular" errors myself.

I'm for it, though I find mechanism for passing multiple errors useful, so we could have some replacement for it. Then we would just have to convert all returned diagnostics into our collection type.

Regarding the code, I would expect to have something like this as an implementation:
```go
func createBackend(logger *logrus.Entry, config *config.Config) backend.Backend {
  bn := config.RootConfig.Backend.Name

  var c interface{}

  switch bn {
  case backend.Local:
    c = &local.Config{}
  case backend.S3:
    c = &s3.Config{}
  }

  if d := gohcl.DecodeBody(config.RootConfig.Backend.Config, config.EvalContext, c); d.HasErrors() {
    logger.Fatalf("fail")
  }

  b, err := c.NewBackend()
  if err != nil {
    logger.Fatalf("fail")
  }

  return b
}
```

So the unique bit per backend is the struct holding the configuration. The rest of the code should be the same for all backends.

Also, I'd really like to not add new code, which use `Fatalf` for flow control - #630. We should be using logger for logging, not for the flow control.

Reply:
Also, what you're proposing kind of breaks encapsulation: now the caller needs to deal with `gohcl.DecodeBody()` which is an implementation detail of the backend AFAICT.

Reply:
>What concrete type does c.NewBackend() return?

It should return `Backend` interface, as in current PR (or specific type implementing backend interface).

>Also, what you're proposing kind of breaks encapsulation: now the caller needs to deal with gohcl.DecodeBody() which is an implementation detail of the backend AFAICT.

This is not the implementation detail of the backend. This is the UI part of the `lokoctl`. The backend should have no knowledge about HCL (maybe except struct tags).

Reply:
>It should return Backend interface, as in current PR (or specific type implementing backend interface).

I'm asking what **concrete** type you expect `NewBackend()` to return. I understand there is `backend.Backend` as the return type in the function's signature but you need to return a struct in the end.

>This is not the implementation detail of the backend. This is the UI part of the lokoctl. The backend should have no knowledge about HCL (maybe except struct tags).

My point is that it doesn't make sense to me to move `gohcl.DecodeBody()` to the UI. Does it make sense to you? Why?

Reply:
>My point is that it doesn't make sense to me to move gohcl.DecodeBody() to the UI. Does it make sense to you? Why?

In my opinion, we should isolate entire HCL code into one place, so it's only job is to convert HCL files into the Go struct, so we can operate on Go structs from there. This way, things are easier to test as test do not require HCL parser.

How about something like the following patch? Though still, in my opinion, having render method on config struct and return an error would even simplify it more.
```patch
From aeb00d3296deacf753ad5a6a33e64301ff51a47c Mon Sep 17 00:00:00 2001
From: Mateusz Gozdek <mateusz@kinvolk.io>
Date: Tue, 25 Aug 2020 18:31:15 +0200
Subject: [PATCH] patch

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>
---
 cli/cmd/cluster.go         | 44 +++++++++++---------------------------
 pkg/backend/backend.go     |  5 +++++
 pkg/backend/local/local.go | 14 ++++++------
 pkg/backend/s3/s3.go       | 12 ++++++-----
 4 files changed, 33 insertions(+), 42 deletions(-)

diff --git a/cli/cmd/cluster.go b/cli/cmd/cluster.go
index dd5d191c..9306905a 100644
--- a/cli/cmd/cluster.go
+++ b/cli/cmd/cluster.go
@@ -17,6 +17,7 @@ package cmd
 import (
        "fmt"

+       "github.com/hashicorp/hcl/v2/gohcl"
        "github.com/mitchellh/go-homedir"
        "github.com/sirupsen/logrus"
        "github.com/spf13/cobra"
@@ -85,44 +86,25 @@ func initialize(ctxLogger *logrus.Entry) (*terraform.Executor, platform.Platform
 func createBackend(logger *logrus.Entry, config *config.Config) backend.Backend {
        bn := config.RootConfig.Backend.Name

+       var backendConfig backend.Config
+
        switch bn {
        case backend.Local:
-               bc, diags := local.NewConfig(&config.RootConfig.Backend.Config, config.EvalContext)
-               if diags.HasErrors() {
-                       for _, diagnostic := range diags {
-                               logger.Error(diagnostic.Error())
-                       }
-
-                       logger.Fatal("Errors found while loading backend configuration")
-               }
-
-               b, err := local.NewBackend(bc)
-               if err != nil {
-                       logger.Fatalf("Error constructing backend: %v", err)
-               }
-
-               return b
+               backendConfig = &local.Config{}
        case backend.S3:
-               bc, diags := s3.NewConfig(&config.RootConfig.Backend.Config, config.EvalContext)
-               if diags.HasErrors() {
-                       for _, diagnostic := range diags {
-                               logger.Error(diagnostic.Error())
-                       }
-
-                       logger.Fatal("Errors found while loading backend configuration")
-               }
-
-               b, err := s3.NewBackend(bc)
-               if err != nil {
-                       logger.Fatalf("Error constructing backend: %v", err)
-               }
+               backendConfig = &s3.Config{}
+       }

-               return b
+       if d := gohcl.DecodeBody(config.RootConfig.Backend.Config, config.EvalContext, backendConfig); d.HasErrors() {
+               logger.Fatalf("bad backend")
        }

-       logger.Fatalf("Unknown backend %q", bn)
+       b, err := backendConfig.New()
+       if err != nil {
+               logger.Fatalf("bad backend")
+       }

-       return nil
+       return b
 }

 // initializeTerraform initialized Terraform directory using given backend and platform
diff --git a/pkg/backend/backend.go b/pkg/backend/backend.go
index 7136ae6b..f21e31da 100644
--- a/pkg/backend/backend.go
+++ b/pkg/backend/backend.go
@@ -26,3 +26,8 @@ type Backend interface {
        // String returns the Backend's configuration as rendered Terraform code.
        String() string
 }
+
+// Config describes functionality of backend configuration.
+type Config interface {
+       New() (Backend, error)
+}
diff --git a/pkg/backend/local/local.go b/pkg/backend/local/local.go
index 4af828a2..bb472926 100644
--- a/pkg/backend/local/local.go
+++ b/pkg/backend/local/local.go
@@ -19,7 +19,9 @@ import (

        "github.com/hashicorp/hcl/v2"
        "github.com/hashicorp/hcl/v2/gohcl"
+
        "github.com/kinvolk/lokomotive/internal/template"
+       "github.com/kinvolk/lokomotive/pkg/backend"
 )

 // Config represents the configuration of a local backend.
@@ -42,23 +44,23 @@ func NewConfig(b *hcl.Body, ctx *hcl.EvalContext) (*Config, hcl.Diagnostics) {
        return c, hcl.Diagnostics{}
 }

-// Backend implements the Backend interface for a local backend.
-type Backend struct {
+// backend implements the Backend interface for a local backend.
+type localBackend struct {
        config *Config
        // A string containing the rendered Terraform code of the backend.
        rendered string
 }

-func (b *Backend) String() string {
+func (b *localBackend) String() string {
        return b.rendered
 }

-// NewBackend constructs a Backend based on the provided config and returns a pointer to it.
-func NewBackend(c *Config) (*Backend, error) {
+// New constructs a Backend based on the provided config and returns a pointer to it.
+func (c *Config) New() (backend.Backend, error) {
        rendered, err := template.Render(backendConfigTmpl, c)
        if err != nil {
                return nil, fmt.Errorf("rendering backend: %v", err)
        }

-       return &Backend{config: c, rendered: rendered}, nil
+       return &localBackend{config: c, rendered: rendered}, nil
 }
diff --git a/pkg/backend/s3/s3.go b/pkg/backend/s3/s3.go
index 8ab4b2fd..7b450615 100644
--- a/pkg/backend/s3/s3.go
+++ b/pkg/backend/s3/s3.go
@@ -19,7 +19,9 @@ import (

        "github.com/hashicorp/hcl/v2"
        "github.com/hashicorp/hcl/v2/gohcl"
+
        "github.com/kinvolk/lokomotive/internal/template"
+       "github.com/kinvolk/lokomotive/pkg/backend"
 )

 // Config represents the configuration of an S3 backend.
@@ -67,18 +69,18 @@ func NewConfig(b *hcl.Body, ctx *hcl.EvalContext) (*Config, hcl.Diagnostics) {
 }

 // Backend implements the Backend interface for an S3 backend.
-type Backend struct {
+type s3Backend struct {
        config *Config
        // A string containing the rendered Terraform code of the backend.
        rendered string
 }

-func (b *Backend) String() string {
+func (b *s3Backend) String() string {
        return b.rendered
 }

-// NewBackend constructs a Backend based on the provided config and returns a pointer to it.
-func NewBackend(c *Config) (*Backend, error) {
+// New constructs a Backend based on the provided config and returns a pointer to it.
+func (c *Config) New() (backend.Backend, error) {
        if err := c.validate(); err != nil {
                return nil, fmt.Errorf("validating backend config: %w", err)
        }
@@ -88,5 +90,5 @@ func NewBackend(c *Config) (*Backend, error) {
                return nil, fmt.Errorf("rendering backend: %v", err)
        }

-       return &Backend{config: c, rendered: rendered}, nil
+       return &s3Backend{config: c, rendered: rendered}, nil
 }
--
2.28.0

```